From dea52ae3e14f3f87fc301dcb3f05f1bdb29091ea Mon Sep 17 00:00:00 2001
From: Daniel Baluta <daniel.baluta@nxp.com>
Date: Thu, 20 Apr 2017 18:27:42 +0300
Subject: [PATCH 1719/5242] MLK-14536: ASoC: wm8960: Fix playback in CPU DAI
 master mode

commit  174c20dae01120e3b0ccbec1bf51c678e5952486 from
https://source.codeaurora.org/external/imx/linux-imx.git

With the current rates for MCLK is not possible to derive bitclk
for all files in S20_3LE format and also for files with S24_LE sampled
at 48000Hz.

In order to fix this, we need to find a better MCLK value. We did this
in two steps:
	1) Use params_physical_width to get rid of S20_3LE burden.
	2) Brute force into all available rates which can pass fsl_sai_set_bclk
	   algorithm.

Thus we found 36864000 to be the smallest acceptable rate for MCLK.

Reviewed-by: Mihai Serban <mihai.serban@nxp.com>
Suggested-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/boot/dts/imx6ul-14x14-evk.dtsi |    4 ++--
 sound/soc/fsl/imx-wm8960.c              |    3 ++-
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/arch/arm/boot/dts/imx6ul-14x14-evk.dtsi b/arch/arm/boot/dts/imx6ul-14x14-evk.dtsi
index 6e23a86..807c67d 100644
--- a/arch/arm/boot/dts/imx6ul-14x14-evk.dtsi
+++ b/arch/arm/boot/dts/imx6ul-14x14-evk.dtsi
@@ -147,7 +147,7 @@
 
 &clks {
 	assigned-clocks = <&clks IMX6UL_CLK_PLL4_AUDIO_DIV>;
-	assigned-clock-rates = <786432000>;
+	assigned-clock-rates = <1179648000>;
 };
 
 &cpu0 {
@@ -339,7 +339,7 @@
 	assigned-clocks = <&clks IMX6UL_CLK_SAI2_SEL>,
 			  <&clks IMX6UL_CLK_SAI2>;
 	assigned-clock-parents = <&clks IMX6UL_CLK_PLL4_AUDIO_DIV>;
-	assigned-clock-rates = <0>, <12288000>;
+	assigned-clock-rates = <0>, <36864000>;
 	fsl,sai-mclk-direction-output;
 	status = "okay";
 };
diff --git a/sound/soc/fsl/imx-wm8960.c b/sound/soc/fsl/imx-wm8960.c
index 7ed03fb..bff066f 100644
--- a/sound/soc/fsl/imx-wm8960.c
+++ b/sound/soc/fsl/imx-wm8960.c
@@ -225,7 +225,8 @@ static int imx_hifi_hw_params(struct snd_pcm_substream *substream,
 	}
 
 	if (!data->is_codec_master) {
-		ret = snd_soc_dai_set_tdm_slot(cpu_dai, 0, 0, 2, params_width(params));
+		ret = snd_soc_dai_set_tdm_slot(cpu_dai, 0, 0, 2,
+					       params_physical_width(params));
 		if (ret) {
 			dev_err(dev, "failed to set cpu dai tdm slot: %d\n", ret);
 			return ret;
-- 
1.7.9.5

